atlasdb-commons/src/main/java/com/palantir/common/concurrent/ConcurrentStreams.java

```diff
@@ -0,0 +1,115 @@
+ */
+
+package com.palantir.common.concurrent;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.palantir.common.base.Throwables;
+
+public class ConcurrentStreams {
+
+    private ConcurrentStreams() {}
+
+    /**
+     * Runs a map function over all elements in a list with a provided executor and concurrency level.
+     *
+     * @param values The elements to be mapped.
+     * @param mapper The function that maps the elements.
+     * @param executor The executor that runs the concurrent operations.
+     * @param concurrency The max number of operations to be run in parallel. Note that this will
+     *        not check the size of the underlying executor, so ideally the executor should have at least
+     *        as many threads as this value.
+     * @return a stream of mapped elements from the provided list.
+     */
+    public static <T, S> Stream<S> map(
+            List<T> values, Function<T, S> mapper, Executor executor, int concurrency) {
+
+        int size = values.size();
+        if (size <= 1 || concurrency == 1) {
+            return values.stream().map(mapper);
+        }
+
+        Map<T, CompletableFuture<S>> futuresByValue = values.stream()
+                .collect(Collectors.toMap(Function.identity(), value -> new CompletableFuture<>()));
+        Queue<T> valueQueue = new ConcurrentLinkedQueue<T>(values);
+
+        if (size < concurrency) {
+            concurrency = size;
+        }
+        for (int i = 0; i < concurrency; i++) {
+            executor.execute(() -> runOperationsAndUpdateFutures(futuresByValue, valueQueue, mapper));
+        }
+        return streamAllUnchecked(futuresByValue.values());
+    }
+
+    private static <T, S> void runOperationsAndUpdateFutures(
+            Map<T, CompletableFuture<S>> futuresByValue, Queue<T> valueQueue, Function<T, S> operation) {
+
+        runUntilEmpty(valueQueue, value -> {
+            CompletableFuture<S> future = futuresByValue.get(value);
+            try {
+                future.complete(operation.apply(value));
+            } catch (Exception e) {
+                future.completeExceptionally(e);
+                completeAllExceptionally(futuresByValue, valueQueue);
+            }
+        });
+    }
+
+    private static <T, S> void completeAllExceptionally(
+            Map<T, CompletableFuture<S>> futuresByValue, Queue<T> valueQueue) {
+
+        runUntilEmpty(valueQueue, value -> futuresByValue.get(value).completeExceptionally(
+                new RuntimeException("Task terminated early due to another concurrent request failing")));
```

Comment:
Should we chain the originating exception onto this or would that just be bloat? Could also preallocate a dummy exception with no stacktrace to avoid log bloat and overhead in times of strife

Reply:
Think more bloat since this is mostly just to guarantee that no one is blocking on those futures and most cases they'll never actually be checked since we're processing the values in-order. So went with the dummy exception approach.